lib/improver/weights.py

```diff
@@ -30,123 +30,259 @@
+                         Options are
+                         evenly - adding the weights from the
+                                   missing forecasts evenly across
+                                   the remaining forecasts.
+                         proportional - re-weight according to the
+                                   proportion of the previous weights.
+                         Default is evenly
+            Returns:
+                redistributed_weights : (numpy.ndarray) of weights
+                                       where sum = 1.0
+                                       and missing weights are set to -1.0
+        """
+        sumval = weights.sum()
+
+        if abs(sumval - 1.0) > 0.0001:
+            msg = 'Sum of weights must be 1.0'
+            raise ValueError(msg)
+
+        if weights.min() < 0.0:
+            msg = 'Weights should be positive or at least one > 0.0'
+            raise ValueError(msg)
+
+        if len(weights) != len(missing_weights):
+            msg = ('Arrays weights and missing_weights not the same size'
+                   ' weights is len {0:}'.format(len(weights)) +
+                   ' missing_weights is len {0:}'.format(len(missing_weights)))
+            raise ValueError(msg)
+
+        missing_sum = missing_weights.sum()
+        if missing_sum > 0.0:
+            combined_weights = weights*missing_weights
+            if method == 'evenly':
+                missing_avg_weight = (1.0 - combined_weights.sum())/missing_sum
+                redistributed_weights = combined_weights + missing_avg_weight
+            elif method == 'proportional':
+                redistributed_weights = (
+                    WeightsUtilities.normalise_weights(combined_weights))
+            else:
+                msg = ('Unknown weights redistribution method'
+                       ': {}'.format(method))
+                raise ValueError(msg)
+            # Set missing values to -1.0
+            redistributed_weights[np.where(missing_weights == 0.0)] = -1.0
         else:
-            slope = (ynval - y0val)/(num_of_weights - 1.0)
+            redistributed_weights = weights
+        return redistributed_weights
+
+    @staticmethod
+    def nonlinear_weights(num_of_weights, cval):
+        """Create nonlinear weights.
+
+            Args:
+                num_of_weights : Positive Integer
+                                 Number of weights to create.
+
+                cval : Float
+                       greater than 0.0 but less than or equal to 1,0,
+                       to be used for the nonlinear weights function.
+                       1.0 = equal weights for all.
+
+                       Weights will be calculated as
+                         cval**(tval-1)/Sum(of all weights)
+                       tval is the value of 1 to num_of_weights,
+
+            Returns:
+                weights : array of weights, sum of all weights = 1.0
+
+        """
```

Comment:
The check on `num_of_weights` below also appears in nonlinear_weights and could be moved to the end of the process_coord method. This would also reduce duplication in the unit tests.